ui/components/app/token-list/token-list.js

```diff
@@ -46,8 +46,18 @@ export default function TokenList({ onTokenClick }) {
   return (
     <div>
       {tokensWithBalances.map((tokenData, index) => {
+        const { isERC721 } = tokens.find(
```

Comment:
Maybe we should update `useTokenTracker` so that `isERC721` is a property on the items in the `tokensWithBalances` array? Then we don't need to search tokens for a matching address. This might make the error I found go away too.

Reply:
So it seems that error can be resolved by switching away from destructuring to optional chaining to protect against an undefined value. But I'll look into using the useTokenTracker as a more elegant solution!

Reply:
@danjm looks like you read [this conversation](https://github.com/MetaMask/metamask-extension/pull/11210#discussion_r653040540). Do you think that we should move from updating token types when users attempt to send to a more pro-active pattern of updating when all tokens in the token list that don't already have the isERC721 flag already set?

Reply:
> Then we don't need to search tokens for a matching address. 

I recognize that for a user with a lot of tokens this adds some overhead for just rendering their token list...

Reply:
Oh wait I see. You're not recommending that we do the update here just add the isERC721 flag where available in the useTokenTracker hook. Disregard my previous comments.

Reply:
@danjm I've opened a PR to add the isERC721 field to the eth-token-tracker class: https://github.com/MetaMask/eth-token-tracker/pull/62. Though I can't seem to get the unit tests to pass

Reply:
@danjm just wanted to check if you've gotten a chance to read through this yet

Reply:
So I agree that this is cleaner from the perspective of extension code. The only challenge with this is that `eth-token-tracker` wasn't really meant to support non-ERC20 tokens. It creates a contract object using https://www.npmjs.com/package/human-standard-token-abi, which is meant for ERC-20. The fact that it works with ERC-721 is sort of an accident, and part of the reason we have this incomplete "support" of ERC-721 in the extension.

For now, I'd rather not add details about ERC-721 at the level of this library. We will need to deal with this soon, when we add actual NFT support. But it might be best to first just band-aid something in the hook, and then we can deal with a broader refactor of eth-token-tracker at a time when we are ready to add proper ERC-721 support.

Sorry! I realize this means redoing your work. I'm open to alternative opinions on this.

Reply:
@danjm no worries! I'll add the band-aid loop in the hook for now!